usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1110 @@
+		sep->se_policy = policy;
+	} else if (*sep->se_policy == '\0') {
+		/* IPsec was intentionally disabled. */
+		free(sep->se_policy);
+		sep->se_policy = NULL;
+	}
+}
+#endif
+
+static void
+try_infer_socktype(struct servtab *sep) {
+	if (sep->se_socktype != SERVTAB_UNSPEC_VAL || sep->se_proto == NULL) {
+		return;
+	}
+
+	/* Check values of se_proto udp, udp6, tcp, tcp6 to set dgram/stream */
+	if (strncmp(sep->se_proto, "udp", 3) == 0) {
+		sep->se_socktype = SOCK_DGRAM;
+	} else if (strncmp(sep->se_proto, "tcp", 3) == 0) {
+		sep->se_socktype = SOCK_STREAM;
+	}
+}
+
+static bool
+setup_internal(struct servtab *sep)
+{
+	pid_t wait_prev = sep->se_wait;
+	if (parse_server(sep, "internal") != 0) {
+		ENI("exec");
+		return false;
+	}
+	
+	if (wait_prev != SERVTAB_UNSPEC_VAL && wait_prev != sep->se_wait) {
+		/* If wait was already specified throw an error. */
+		WRN(WAIT_WRN, sep->se_service, (sep->se_wait ? "yes" : "no"));
+	}
+	return true;
+}
+
+static bool
+infer_protocol_ip_version(struct servtab *sep)
+{
+	struct in_addr tmp;
+
+	if (strcmp("tcp", sep->se_proto) != 0
+		&& strcmp("udp", sep->se_proto) != 0
+		&& strcmp("rpc/tcp", sep->se_proto) != 0
+		&& strcmp("rpc/udp", sep->se_proto) != 0) {
+		return true;
+	}
+	
+	if (inet_pton(AF_INET, sep->se_hostaddr, &tmp)) {
+		sep->se_family = AF_INET;
+		return true;
+	}
+
+	if (inet_pton(AF_INET6, sep->se_hostaddr, &tmp)) {
+		sep->se_family = AF_INET6;
+		return true;
+	}
+
+	ERR("Address family of %s is ambigous or invalid. "
+		"Explicitly specify protocol", sep->se_hostaddr);
+	return false;
+}
+
+/* 
+ * Skips whitespaces, newline characters, and comments, 
+ * and returns the next token. Returns false and logs error if an EOF is 
+ * encountered. 
+ */
+static bool 
+skip_whitespace(char **cpp) {
+	char *cp = *cpp;
+
+	int line_start = line_number;
+	
+	for (;;) {
+		while (isblank((unsigned char)*cp))
+			cp++;
+
+		if (*cp == '\0' || *cp == '#') {  
+			cp = nextline(fconfig);
+
+			/* Should never expect EOF when skipping whitespace */
+			if (cp == NULL) {
+				ERR("Early end of file after line %d", line_start); 
+				return false;
+			}
+			continue;
+		}
+		break;
+	}
+
+	*cpp = cp;
+	return true;
+}
+
+/* Get the key handler function pointer for the given name */
+static key_handler_func 
+get_handler(char *name) {
+	/* Call function to handle option parsing. */
+	for(size_t i = 0; i < __arraycount(key_handlers); i++) {
+		if (strcmp(key_handlers[i].name, name) == 0) {
+			return key_handlers[i].handler;
+		}
+	}
+	return NULL;
+}
+
+/* 
+ * Perform an in-place parse of a single-line quoted string 
+ * with escape sequences. Sets *cpp to the position after the quoted characters.
+ * Uses shell-style quote parsing.
+ */
+static bool 
+parse_quotes(char **cpp) {
+
+	char *cp = *cpp;
+	char quote = *cp;
+
+	memmove(cp, cp+1, strlen(cp+1)+1); 	
+	
+	while (*cp && quote) {
+		if (*cp == quote) {
+			quote = '\0';
+			memmove(cp, cp+1, strlen(cp+1)+1);
```

Comment:
I think this memmove pattern is used too frequently and could be made into an inline function:
```
static inline void 
strmove(char *buf, size_t off) { 
            memcpy(buf, buf + off, strlen(buf + off) + 1);
}